src/phoenix/scheduler/task_scheduler.go

```diff
@@ -347,59 +374,37 @@ func (ts *TaskScheduler) enqueueJob(enqueueCount int, jobId string) error {
 		}
 		queuePos := 0
 
-		targetWorkerId := probeNodesList[targetIndex%len(probeNodesList)]
-
 		ts.workerLock.Lock()
-		// fmt.Println("[TaskScheduler: enqueueJob] workerLock acquired")
+
+		// relax constraint that we send RPCs to distinct machines for the sake of concurrency and speed
+
+		// dynamically choose workerId. While holding onto the lock, make an RPC call to the client
+		targetWorkerId := ts.workerIds[rand.Intn(len(ts.workerIds))]
 		targetMonitor := ts.MonitorClientPool[targetWorkerId]
-		// fmt.Println("[TaskScheduler: enqueueJob] workerLock Unlocked")
 		ts.workerLock.Unlock()
 
 		if e := targetMonitor.EnqueueReservation(taskR, &queuePos); e != nil {
 			fmt.Printf("[TaskScheduler %s: enqueueJob]: Failed to enqueue reservation on %s: %v\n",
 				ts.Addr, targetWorkerId, e)
+			continue
 		}
 
-		fmt.Printf("[TaskScheduler %s: enqueueJob]: Enqueuing reservation on monitor %d for job reservation %s\n",
-			ts.Addr, targetWorkerId, taskR.JobID)
+		ts.workerLock.Lock()
+		_, exists := ts.workerIdToJobReservations[targetMonitor.Addr]
+		if ! exists {
+			ts.workerIdToJobReservations[targetMonitor.Addr] = make(map[string]int)
+		}
 
-		// if e != nil {
-		// 	// Remove the inactive back
-		// 	probeNodesList = append(probeNodesList[:targetIndex], probeNodesList[targetIndex+1:]...)
-		// 	continue
-		// }
+		// targetMonitor.Addr has one more jobId in it's queue
+		ts.workerIdToJobReservations[targetMonitor.Addr][jobId] ++
+		ts.workerLock.Unlock()
+
+		fmt.Printf("[TaskScheduler %s: enqueueJob]: Enqueuing reservation on monitor %s for job reservation %s\n",
+			ts.Addr, targetWorkerId, taskR.JobID)
 
 		enqueueCount--
-		targetIndex++
 	}
 
 	return nil
 }
 
-func (ts *TaskScheduler) selectEnqueueWorker(probeCount int) []string {
```

Comment:
We relax the constraint that we always send enqueue reservations to `min(m*d, # workers)` distinct workers to distribute jobs faster and be prevent problems related to stale data. One problem that has occurred in the past with usage of `selectEnqueueWorker` is that the list of valid workerAddresses generated by `selectEnqueueWorker` is stale by the time we perform a for loop in `enqueueJob` to send reservation requests. If a node dies after the list is generated and before we send an enqueueReservation to that node, the enqueueReservation is lost.